hazelcast/src/main/java/com/hazelcast/config/YamlConfigBuilder.java

```diff
@@ -0,0 +1,302 @@
+        replaceVariables(w3cRootNode);
+        importDocuments(imdgRoot);
+
+        new YamlMemberDomConfigProcessor(true, config).buildConfig(w3cRootNode);
+    }
+
+    private void importDocuments(YamlNode imdgRoot) throws Exception {
+        YamlMapping rootAsMapping = asMapping(imdgRoot);
+        YamlSequence importSeq = rootAsMapping.childAsSequence(ConfigSections.IMPORT.name);
+        if (importSeq == null || importSeq.childCount() == 0) {
+            return;
+        }
+
+        for (YamlNode importNode : importSeq.children()) {
+            String resource = asScalar(importNode).nodeValue();
+            URL url = ConfigLoader.locateConfig(resource);
+            if (url == null) {
+                throw new InvalidConfigurationException("Failed to load resource: " + resource);
+            }
+            if (!currentlyImportedFiles.add(url.getPath())) {
+                throw new InvalidConfigurationException("Cyclic loading of resource '" + url.getPath() + "' detected!");
+            }
+
+            YamlNode rootLoaded;
+            try {
+                rootLoaded = YamlLoader.load(url.openStream());
+            } catch (Exception ex) {
+                throw new InvalidConfigurationException("Loading YAML document from resource " + url.getPath() + " failed", ex);
+            }
+            YamlNode imdgRootLoaded = asMapping(rootLoaded).child(ConfigSections.HAZELCAST.name().toLowerCase());
+
+            replaceVariables(asW3cNode(imdgRootLoaded));
+            importDocuments(imdgRootLoaded);
+
+            // we need to merge and not just substitute with the content of the imported document
+            // YAML documents define mappings where the name of the nodes should be unique
+            merge(imdgRootLoaded, imdgRoot);
+        }
+
+        ((MutableYamlMapping) rootAsMapping).removeChild(ConfigSections.IMPORT.name);
+    }
+
+    /**
+     * Merges the source YAML document into the target YAML document
+     * <p/>
+     * If a given source node is not found in the target, it will be attached
+     * If a given source node is found in the target, this method is invoked
+     * recursively with the given node
+     *
+     * @param source The source YAML document's root
+     * @param target The target YAML document's root
+     */
+    private void merge(YamlNode source, YamlNode target) {
+        if (source == null) {
+            return;
+        }
+
+        YamlMapping sourceAsMapping = asMapping(source);
+        YamlMapping targetAsMapping = asMapping(target);
+
+        for (YamlNode sourceChild : sourceAsMapping.children()) {
+            YamlNode targetChild = targetAsMapping.child(sourceChild.nodeName());
+            if (targetChild != null) {
+                merge(sourceChild, targetChild);
+            } else {
+                if (targetAsMapping instanceof MutableYamlMapping) {
+                    ((YamlMappingImpl) targetAsMapping).addChild(sourceChild.nodeName(), sourceChild);
```

Comment:
Should we just always cast to `MutableYamlMapping` without `instanceof`?

Reply:
:+1: Done.